### PR TITLE
[GEOS-10469] Make the request logger configurable from the UI

### PR DESCRIPTION
**Pull request by turingtestfail**
[GEOS-10469](https://osgeo-org.atlassian.net/browse/GEOS-10469)

started on populating ui and configuration state objects

Got the enable part working

Got the headers and body check boxes working and enabled depending on main enabled

change layout so checkboxes are hierarchical

added logging filter unit test and null check

cleanup

Fixed test so that we ensure logging object is consistent

documentation

  

# Checklist

- [x] I have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/main/CONTRIBUTING.md).
- [x] I have sent a [Contribution Licence Agreement](https://docs.geoserver.org/latest/en/developer/policies/committing.html) (not required for small changes, e.g., fixing typos in documentation).
- [x] First PR targets the `main` branch (backports managed later; ignore for branch specific issues).
- [x] All the build checks are green ([see automated QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html)).

For core and extension modules:

- [x] New unit tests have been added covering the changes.
- [x] [Documentation](https://github.com/geoserver/geoserver/tree/main/doc/en/user/source) has been updated (if change is visible to end users).
- [ ] The [REST API docs](https://github.com/geoserver/geoserver/tree/main/doc/en/api/1.0.0) have been updated (when changing configuration objects or the REST controllers).
- [x] There is an issue in the [GeoServer Jira](https://osgeo-org.atlassian.net/browse/GEOS/summary) (except for changes that do not affect administrators or end users in any way).
- [x] Commit message(s) must be in the form ``[GEOS-XYZWV] Title of the Jira ticket``.
- [x] Bug fixes and small new features are presented as a single commit.
- [x] Each commit has a single objective (if there are multiple commits, each has a separate JIRA ticket describing its goal).

### CODE REVIEW

**Comment by jodygarnett** [comment]
Please hold off on this commit pending https://github.com/geoserver/geoserver/pull/5823

I did add some logic here to avoid logging body when application/zip (and other binary content).

File: src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java
Comment (jodygarnett) [inline src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java:9]: Okay, so this is a public class and will need javadocs - but more importantly it is a change to configuration on disk and will need a change proposal.

For starts I would double check if you can add the extra isRequestLoggingEnabled and isRequestBodiesEnabled to global settings, there is a metadata map which is ... specifically done so you can add extra settings like this without having to make a change proposal.

**Review by jodygarnett — CHANGES_REQUESTED** [review]
Please review GlobalSettings metadata map to see if your two enabled flags can be handled without configuration change.

a) If this is the case your PR can be modified to not require a change proposal.
b) If you need an additional object, then it may be worth isolating all the logging details into one catalog info (as there is already some difficulty with reading this information from disk before spring-framework is properly available).

I ask that this PR be held back pending https://github.com/geoserver/geoserver/pull/5823 which also affects logging when/how header bodies are logged.

File: src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java:9]: Thanks Jody.  It would make more sense to piggyback off of an existing settings model.  When you refer to a "metadata map" are you talking about the MetadataMap called "metadata" within GeoServerInfoImpl?

Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java:9]: Or are you referring in general to the globalInfoModel on the GlobalSettingsPage?

Comment (jodygarnett) [reply src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java:9]: I that sounds about right

File: doc/en/user/source/production/troubleshooting.rst
Comment (jodygarnett) [reply doc/en/user/source/production/troubleshooting.rst:28]: I am afraid you will need to recreate these screen snaps and so forth after #5823 is merged; best to hold off for now; or try rebasing your work against the log4j2_upgrade branch so it is ready to go when the time comes?

File: src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/config/RequestLoggingInfo.java]: Fixed to use MetadataMap

**Comment by jodygarnett** [comment]
Just an idea, do you want to change your PR to be against the log4j2_upgrade branch?

**Comment by aaime** [comment]
@jodygarnett we are making the documentation change avoid sharing the same picture, making the two PRs independent of each other.

File: doc/en/user/source/production/troubleshooting.rst
Comment (turingtestfail) [reply doc/en/user/source/production/troubleshooting.rst:28]: screencap reverted

**Review by aaime — CHANGES_REQUESTED** [review]
@turingtestfail found the reason for the glitchy behavior in the global page, about to push a change, was too hard to describe in english...

File: src/web/core/src/main/java/org/geoserver/web/admin/ServerAdminPage.java
Comment (aaime) [inline src/web/core/src/main/java/org/geoserver/web/admin/ServerAdminPage.java:34]: This was the source of the glitches, ajax calls made the loadable detachable model reload the config object and changes would go bye-bye.

**Comment by aaime** [comment]
@turingtestfail can you counter-review the change I've just pushed?

**Comment by jodygarnett** [comment]
Thank you for waiting, you should now be able to rebase and update the global settings page and documentation without trouble.

File: doc/en/user/source/production/troubleshooting.rst
Comment (turingtestfail) [reply doc/en/user/source/production/troubleshooting.rst:28]: rebased

File: doc/en/user/source/production/troubleshooting.rst
Comment (turingtestfail) [reply doc/en/user/source/production/troubleshooting.rst:30]: switched to other screenshot

File: src/main/src/main/java/org/geoserver/filters/LoggingFilter.java
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/filters/LoggingFilter.java]: used subclass as recommended

File: src/main/src/main/java/org/geoserver/filters/LoggingFilter.java
Comment (turingtestfail) [reply src/main/src/main/java/org/geoserver/filters/LoggingFilter.java]: assigned to local variables

File: src/web/core/src/main/java/org/geoserver/web/admin/ServerAdminPage.java
Comment (turingtestfail) [reply src/web/core/src/main/java/org/geoserver/web/admin/ServerAdminPage.java:34]: reviewed

File: src/main/src/test/java/org/geoserver/filters/LoggingFilterTest.java
Comment (turingtestfail) [reply src/main/src/test/java/org/geoserver/filters/LoggingFilterTest.java]: method factored out and switched to use strings

**Comment by aaime** [comment]
JDK 17 failure is unrelated, being addressed on https://github.com/geoserver/geoserver/pull/5856. 
Merging.

**Comment by aaime** [comment]
@jodygarnett seems you already cut 2.21.x, I guess we're going to wait a little for a backport (we wanted to do an early backport, based on how useful this is for day to day issue debugging, and the fact it's introducing no new mandatory configuration elements).

**Comment by jodygarnett** [comment]
I only cut the RC yesterday, I made the forks a couple days ago but it took some time to trouble shoot builds.

If you wish merge now, and backport, I can run the job again before monday anouncement.

**Comment by jodygarnett** [comment]
backport to 2.21.x and remaking 2.21-RC.

I note the above checklist does not update the REST API, I think that is okay since it is "only" another addition to the metadata map...

**Comment by jodygarnett** [comment]
In reviewing the docs I note that there are two ways to disable logging of request bodies
image

Set the number of characters to log for incoming POST requests to 0, or disable Log Request Bodies?

Is that the case? Should ``Number of characters to log for incoming POST requests`` be disabled if Log Request Bodies is disabled?

**Comment by turingtestfail** [comment]
It looks like the number of characters setting does not impact the logging
that occurs with "Log Request Bodies".  Number of characters is associated
the Dispatcher in gs-ows, which has its own logging mechanism.  I can see
how this may be a source of confusion for users.  Maybe the best approach
would be to add notes to the documentation explaining that they are
independent?

**Comment by turingtestfail** [comment]
@jodygarnett In the interests of functionality consolidation and user clarity and to avoid memory issues with the request logging, what do you think about removing the logging from the gs-ows Dispatcher and instead applying the "Number of characters..." parameter to the Request Logging in the LoggingFilter?  I can get started on a PR to that effect right away.

**Comment by jodygarnett** [comment]
@turingtestfail I think that would be a good idea (having duplication in the codebase is no fun). 
You may be able to keep some of the logic to avoid logging out binary content (which was causing trouble for me when testing).

**Pull request merged**
